llvm/lib/Transforms/Utils/LoopIterCountAssumptions.cpp

```diff
@@ -0,0 +1,262 @@
+
+  Value *Cmp = Builder.CreateICmp(Pred, LHS, RHS);
+
+  // Insert Assumption
+  CallInst *Assumption = Builder.CreateAssumption(Cmp);
+  AC.registerAssumption(dyn_cast<AssumeInst>(Assumption));
+  LLVM_DEBUG(dbgs() << "With Comparator             :" << *Cmp << "\n"
+                    << "Assume                      :" << *Assumption << "\n");
+}
+
+/// Determine if the \param CurrentLoop is not rotated yet and Loop Iteration
+/// Count Metadata greater than 0 \return Minimum Iteration Count of the Loop
+std::optional<int64_t> getValidMinIterCount(Loop &CurrentLoop) {
+  BasicBlock *LoopHeader = CurrentLoop.getHeader();
+
+  // Dump loop summary
+  LLVM_DEBUG(if (CurrentLoop.getLoopPreheader()) {
+    dbgs() << "Preheader:" << CurrentLoop.getLoopPreheader()->getName() << "\n";
+  } dbgs() << "LoopIterCountAssumption-Info: Function = "
+           << getFunctionAndBlockNames(*LoopHeader) << "\n");
+
+  std::optional<int64_t> RawMinIterationCount = getMinTripCount(&CurrentLoop);
+  if (!RawMinIterationCount) {
+    LLVM_DEBUG(dbgs() << "LoopIterCountAssumptions: Loop Iteration "
+                         "Count not provided for "
+                      << getFunctionAndBlockNames(*LoopHeader) << "\n");
+    return std::nullopt;
+  }
+
+  const int64_t MinIterCount = *RawMinIterationCount;
+  if (MinIterCount <= 0) {
+    LLVM_DEBUG(dbgs() << "LoopIterCountAssumptions-Warning: Loop Iteration "
+                         "Count is smaller or equal to zero for "
+                      << getFunctionAndBlockNames(*LoopHeader) << "\n");
+    return std::nullopt;
+  }
+
+  LLVM_DEBUG(dbgs() << "Processing Loop Iteration Count Metadata: "
+                    << getFunctionAndBlockNames(*LoopHeader) << " ("
+                    << MinIterCount << ")\n");
+  return MinIterCount;
+}
+
+void insertIterationAssumptions(ICmpInst &LoopCmpInstr, Loop &CurrentLoop,
+                                int64_t IterCount,
+                                LoopStandardAnalysisResults &AR) {
+  const bool ContainsEqualPredicate =
+      LoopCmpInstr.getPredicate() == CmpInst::ICMP_EQ ||
+      LoopCmpInstr.getPredicate() == CmpInst::ICMP_NE;
+
+  // Guarantee that the Loop will execute at least once, to handle variable
+  // StepSizes and EQ/NE predicates
+  if (hasVariableStepSize(*LoopCmpInstr.getOperand(0), AR.SE) ||
+      hasVariableStepSize(*LoopCmpInstr.getOperand(1), AR.SE) ||
+      ContainsEqualPredicate)
+    insertIterationAssumption(LoopCmpInstr, CurrentLoop, 0, AR.SE, AR.AC,
+                              &AR.DT);
+
+  // Do not insert another Assumption with IterCount with EQ or
+  // NE Predicates, since they do not work well with the currently used
+  // assumption-construction-method of reusing the LoopCmpInstr Predicate
+  if (ContainsEqualPredicate) {
+    LLVM_DEBUG(dbgs() << "LoopIterCountAssumptions-Warning: IterCount "
+                         "Assumption Insertion for EQ and NE no supported!");
+    return;
```

Comment:
is that dead code?

Reply:
No, we have to abort the Insertion of Assumptions at this point.
Since we cannot make an assumption at `IterCount` with our current methodology when we have a `NE/EQ branch condition`, we can only guarantee that we will enter the loop once.

FYI: If we have multiple assumptions, only the last one counts and the previous ones will be ignored. Therefore, it would be really bad to add another assumption at this point.

Reply:
Yes I missed that the previous `if` isn't an early return. Does that mean that for predicates other than NE/EQ, we might insert two types of assumptions? How does `hasVariableStepSize` work actually? Won't `insertIterationAssumption` bail out because it can find neither an AddRec nor a loop-invariant value?

Reply:
Oh I think I start to understand, the operand could be a `SCEVAddRecExpr` of degree `>1`, in which case the step is itself a `SCEVAddRecExpr`. But still, why does that need to be handled specifically? Why not just evaluate the SCEV at iteration `MinIterCount - 1`?

Reply:
But it is also not really useful, since e.g.

`N != 0` and `N != 4` does not guarantee us, that N can't be `1,2,3` which limits the expressiveness of the second assumption. We would have to add assumptions for every single Iteration evaluation to get the same expressiveness, which may explode the code a little, if `IterCount` gets high.

Reply:
yes we insert two assumptions. with variable StepSizes, `AddRec` is actually found and the `Stepsize` is now a `Value` instead of a `Constant`.

Consider the following:

```
for (int i =0; i < n; i+=m) {
}
```

If we only add `MinIterCount - 1` assumption, we only guarantee, that `m* (MinIterCount-1) < n` , but this does not guarantee, that we will actually enter the loop. Consider m to be negative. We then guarantee, that n is negative. But we don't actually guarantee, that we enter the loop the first time. 
Therefore we also have to guarantee that by also adding `0 < n`.

With NE this is even more clear

```
for (int i=0 ; i != n ; i++) {
}
```

`MinIterCount-1 != n`, does not actually guarantee, that we will enter the loop, since we don't know if `n != 0` is true